platforms/jvm/toolchains-jvm/src/main/java/org/gradle/jvm/toolchain/internal/install/JdkCacheDirectory.java

```diff
@@ -223,49 +247,34 @@ private static String getInstallFolderName(JvmInstallationMetadata metadata) {
         String os = OperatingSystem.current().getFamilyName();
         return String.format("%s-%s-%s-%s", vendor, version, architecture, os)
                 .replaceAll("[^a-zA-Z0-9\\-]", "_")
-                .toLowerCase();
+                .toLowerCase() + ".2";
     }
 
     private File unpack(File jdkArchive) {
         final FileTree fileTree = asFileTree(jdkArchive);
         String unpackFolderName = getNameWithoutExtension(jdkArchive);
-        final File unpackFolder = new File(jdkDirectory, unpackFolderName);
-        if (!unpackFolder.exists()) {
-            operations.copy(spec -> {
-                spec.from(fileTree);
-                spec.into(unpackFolder);
-                spec.setDuplicatesStrategy(DuplicatesStrategy.WARN);
-            });
-        }
+        final File unpackFolder = temporaryFileProvider.createTemporaryDirectory(unpackFolderName, null, "jdks");
+        unpackFolder.deleteOnExit();
+        operations.copy(spec -> {
+            spec.from(fileTree);
+            spec.into(unpackFolder);
+            spec.setDuplicatesStrategy(DuplicatesStrategy.WARN);
+        });
 
         return unpackFolder;
     }
 
-    private File markedLocation(File unpackFolder) {
-        File[] content = unpackFolder.listFiles();
-        if (content == null) {
-            //can't happen, the installation location is a directory, we have created it
-            throw new RuntimeException("Programming error");
-        }
-
-        //mark the first directory since there should be only one
-        for (File file : content) {
-            if (file.isDirectory()) {
-                return file;
-            }
-        }
-
-        //there were no sub-directories in the installation location
-        return unpackFolder;
-    }
-
     @SuppressWarnings("ResultOfMethodCallIgnored")
-    private File markAsReady(File destination) {
+    private void markAsReady(File root) {
+        try {
+            new File(root, LEGACY_MARKER_FILE).createNewFile();
```

Comment:
💭 Just to confirm, we continue to to create the legacy marker file so that prior gradle versions can use installed toolchains from this gradle version? If so, might be worth adding that as a comment

Reply:
Yes, I will do that.